nevergrad/functions/photonics/photonics.py

```diff
@@ -220,3 +227,240 @@ def morpho(X: np.ndarray) -> float:
+        Type=np.append(0,np.append(np.tile(np.array([1,2]),int(X.size/2)),3))
+        hauteur=np.append(0,np.append(X,30000))
+        pol=0
+        if pol==0:
+            f=Mu
+        else:
+            f=Epsilon
+        k0=2*pi/lam
+        g=Type.size
+        alpha=np.sqrt(Epsilon[Type[0]]*Mu[Type[0]])*k0*sin(theta)
+        gamma=np.sqrt(Epsilon[Type]*Mu[Type]*k0**2-np.ones(g)*alpha**2)
+        if np.real(Epsilon[Type[0]])<0 and np.real(Mu[Type[0]])<0 :
+            gamma[0]=-gamma[0]
+        if g>2:
+            gamma[1:g-2]=gamma[1:g-2]*(1-2*(np.imag(gamma[1:g-2])<0))
+        if np.real(Epsilon[Type[g-1]])<0 and np.real(Mu[Type[g-1]])<0 and np.real(np.sqrt(Epsilon[Type[g-1]]*Mu[Type[g-1]]*k0**2-alpha**2))!=0:
+            gamma[g-1]=-np.sqrt(Epsilon[Type[g-1]]*Mu[Type[g-1]]*k0**2-alpha**2)
+        else :
+            gamma[g-1]=np.sqrt(Epsilon[Type[g-1]]*Mu[Type[g-1]]*k0**2-alpha**2)
+        T=np.zeros(((2*g,2,2)),dtype=complex)
+        T[0]=[[0,1],[1,0]]
+        for k2 in range(g-1):
+            t=np.exp(i*gamma[k2]*hauteur[k2])
+            T[2*k2+1]=[[0,t],[t,0]]
+        #Interface scattering matrix
+            b1=gamma[k2]/f[Type[k2]]
+            b2=gamma[k2+1]/f[Type[k2+1]]
+            T[2*k2+2]=[[(b1-b2)/(b1+b2),2*b2/(b1+b2)],[2*b1/(b1+b2),(b2-b1)/(b1+b2)]]
+            t=np.exp(i*gamma[g-1]*hauteur[g-1])
+            T[2*g-1]=[[0,t],[t,0]]
+        H=np.zeros(((2*g-1,2,2)),dtype=complex)
+        A=np.zeros(((2*g-1,2,2)),dtype=complex)
+        H[0]=T[2*g-1]
+        A[0]=T[0]
+        for j in range(2*g-2):
+            A[j+1]=cascade2(A[j],T[j+1])
+            H[j+1]=cascade2(T[2*g-2-j],H[j])
+        r=A[len(A)-1][0,0]
+        t=A[len(A)-1][1,0]
+        I=np.zeros(((2*g,2,2)),dtype=complex)
+        for j in range(len(T)-1):
+            I[j][0,0]=A[j][1,0]/(1-A[j][1,1]*H[len(T)-2-j][0,0])
+            I[j][0,1]=A[j][1,1]*H[len(T)-2-j][0,1]/(1-A[j][1,1]*H[len(T)-2-j][0,0])
+            I[j][1,0]=A[j][1,0]*H[len(T)-2-j][0,0]/(1-A[j][1,1]*H[len(T)-2-j][0,0])
+            I[j][1,1]=H[len(T)-2-j][0,1]/(1-A[j][1,1]*H[len(T)-2-j][0,0])
+        I[2*g-1][0,0]=I[2*g-2][0,0]*np.exp(i*gamma[g-1]*hauteur[g-1])
+        I[2*g-1][0,1]=I[2*g-2][0,1]*np.exp(i*gamma[g-1]*hauteur[g-1])
+        I[2*g-1][1,0]=0
+        I[2*g-1][1,1]=0
+        w=0
+        poynting=np.zeros(2*g,dtype=complex)
+        if pol==0: #TE
+            for j in range(2*g):
+                poynting[j]=np.real( (I[j][0,0]+I[j][1,0]) * np.conj( (I[j][0,0]-I[j][1,0] )*gamma[w]/Mu[Type[w]] )*Mu[Type[0]]/(gamma[0]))
+                w=w+1-np.mod(j+1,2)
+        else:       #TM
+            for j in range(2*g):
+                poynting[j]=np.real( (I[j][0,0]-I[j][1,0]) * np.conj( (I[j][0,0]+I[j][1,0] )*gamma[w]/Epsilon[Type[w]] )*Epsilon[Type[0]]/(gamma[0]))
+                w=w+1-np.mod(j+1,2)
+        tmp=abs(-np.diff(poynting))
+        absorb=tmp[np.arange(0,2*g,2)]
+        r=A[len(A)-1][0,0]
+        t=A[len(A)-1][1,0]
+        R=abs(r)**2
```

Comment:
this `R` variable is not used, is that expected?

Reply:
Yes it is expected, R is not necessary here, it is here because this code is inspired by another code in which R is usefull.